add support for unmappedTargetPolicy

hduelme opened the pull request:
I added support for `unmappedTargetPolicy`. The policy will be read from the class annotation `@Mapper` first. If not configured on class level, the property of the mapper-config-class will be used. Otherwise the fallback policy warn is used.

To directly match the compiler results, `unmappedTargetPolicy = ERROR` results in Errors getting reported instead of warnings. 
If set to `unmappedTargetPolicy = IGNORE` all warnings for unmapped  targets are suppressed in the current class.

Fixes #132

filiphr commented on src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java:
```diff
@@ -479,4 +482,78 @@ private static Stream<PsiClass> findReferencedMappersOfMapperConfig(PsiAnnotatio
+    }
+
+    @NotNull
+    public static ReportingPolicy getReportingPolicyFromClass( @NotNull PsiClass containingClass,
```
This doesn't need to be public I think

hduelme replied:
Changed to `private`

filiphr commented on src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java:
```diff
@@ -479,4 +482,78 @@ private static Stream<PsiClass> findReferencedMappersOfMapperConfig(PsiAnnotatio
     }
 
+    @NotNull
+    public static ReportingPolicy getReportingPolicyFromMethode( @NotNull PsiMethod method,
```
I would suggest to rename this to `getUnmappedTargetPolicy` and only pass the `mappingMethod`. We are using this method in only one place so we can do the fallback in the implementation of the method and the use of `unmappedTargetPolicy`.

hduelme replied:
Renamed to getUnmappedTargetPolicy

filiphr commented on src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
We don't need to override this. Let's use the default for our tests (Java 1.8)

hduelme replied:
Removed

filiphr commented on src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
We don't need to override this. Let's use the default for our tests (Java 1.8)

hduelme replied:
Removed

filiphr commented on src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyErrorTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
We don't need to override this. Let's use the default for our tests (Java 1.8)

hduelme replied:
Removed

filiphr commented on src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
We don't need to override this. Let's use the default for our tests (Java 1.8)

hduelme replied:
Removed

filiphr commented on src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
We don't need to override this. Let's use the default for our tests (Java 1.8)

hduelme replied:
Removed

filiphr reviewed (changes requested):
Thanks @hduelme.

I've left some comments inline.

Apart from those can we perhaps do the following:

* remove the `_after` tests. I think that it is enough to only validate that the warnings / errors are there.
* Can we please also add some tests where you have `unmappedTargetPolicy = WARN" on the mapper, but `unmappedTargetPolicy = ERROR` on the mapper config?
* Can we please also take `BeanMapping#unmappedTargetPolicy` into consideration as well?

hduelme commented:
@filiphr I added support for `@BeanMapping`. I also renamed the methods as suggested and removed the not needed parameters. Furthermore, I have added more tests, which cover different possible combinations of overwriting policy.

filiphr commented:
I also updated the description of the inspection to include the fact that the unmapped target policy is taken into consideration

hduelme commented:
@filiphr thanks for the feedback. I personally was not sure, if it would be better to have all tests in the same place. But I a agree. It looks cleaner.
